crates/uv-installer/src/compile.rs

```diff
@@ -0,0 +1,207 @@
+use walkdir::WalkDir;
+
+const COMPILEALL_SCRIPT: &str = include_str!("pip_compileall.py");
+const MAIN_TIMEOUT: Duration = Duration::from_secs(10);
+
+#[derive(Debug, Error)]
+pub enum CompileError {
+    #[error("Failed to list files in site-packages")]
+    Walkdir(#[from] walkdir::Error),
+    #[error("Couldn't determine number of cores")]
+    AvailableParallelism(#[source] io::Error),
+    #[error("Failed to send task to worker")]
+    WorkerDisappeared(SendError<PathBuf>),
+    #[error("The task executor is broken, did some other task panic?")]
+    Join(#[from] JoinError),
+    #[error("Failed to start Python interpreter to run compile script")]
+    PythonSubcommand(#[source] io::Error),
+    #[error("Failed to create temporary script file")]
+    TempFile(#[source] io::Error),
+    #[error("Bytecode compilation failed, expected {0:?}, got: {1:?}")]
+    WrongPath(String, String),
+    #[error("Failed to write to Python stdin")]
+    ChildStdin(#[source] io::Error),
+    #[error("Failed to read Python stdout")]
+    ChildStdout(#[source] io::Error),
+    #[error("Bytecode timed out ({}s)", _0.as_secs_f32())]
+    Timeout(Duration),
+}
+
+/// Bytecode compile all file in `dir` using a pool of work-stealing python interpreters running a
+/// Python script that calls `compileall.compile_file`.
+///
+/// All compilation errors are muted (like pip). There is a 10s timeout to catch anything gotten
+/// stuck; This happens way to easily with channels and subprocesses, e.g. because some pipe is
+/// full, we're waiting when it's buffered or we didn't handle channel closing properly.
+///
+/// We only compile all files, but we don't update the RECORD, relying on PEP 491:
+/// > Uninstallers should be smart enough to remove .pyc even if it is not mentioned in RECORD.
+pub async fn compile_tree(dir: &Path, python_executable: &Path) -> Result<usize, CompileError> {
+    let workers =
+        std::thread::available_parallelism().map_err(CompileError::AvailableParallelism)?;
+    // 10 is an arbitrary number.
+    let (sender, receiver) = async_channel::bounded::<PathBuf>(workers.get() * 10);
+
+    // Start the workers.
+    let mut worker_handles = Vec::new();
+    for _ in 0..workers.get() {
+        worker_handles.push(tokio::task::spawn(worker(
+            dir.to_path_buf(),
+            python_executable.to_path_buf(),
+            receiver.clone(),
+        )));
+    }
+
+    // Start the producer, sending all `.py` files to workers.
+    let mut source_files = 0;
+    let mut send_error = None;
+    for entry in WalkDir::new(dir) {
+        let entry = entry?;
+        // https://github.com/pypa/pip/blob/3820b0e52c7fed2b2c43ba731b718f316e6816d1/src/pip/_internal/operations/install/wheel.py#L593-L604
+        if entry.metadata()?.is_file() && entry.path().extension().is_some_and(|ext| ext == "py") {
+            source_files += 1;
+            match tokio::time::timeout(MAIN_TIMEOUT, sender.send(entry.path().to_owned())).await {
```

Comment:
Usually, a channel send timing out should be regarded as a bug. That is, it should never happen. But I think I see why you're doing it here. From what I can tell, the subprocesses themselves might get stuck, and thus if all the workers get stuck, there is nobody to drain from the channel and thus the bounded send blocks when the buffer in the channel becomes full.

I'll suggest some ideas below for how to deal with this.

(It may be the case that we still want to put a timeout here so that we can be extra defensive. I have mixed feelings on that.)